### PR TITLE
add flow types to source maps worker

### PR DESCRIPTION
**Pull request by sankha93**
Associated Issue: #1005
### Summary of Changes
- Type the Source Map Worker
### Testing
- [x] passes `npm test`
- [x] passes `npm run lint`

### CODE REVIEW

File: public/js/utils/path.js [hunk 1]
```diff
@@ -1,20 +1,26 @@
...
   return path.slice(0, idx);
 }
 
-function isURL(str) {
+function isURL(str: string) : boolean {
   return str.indexOf("://") !== -1;
 }
 
-function isAbsolute(str) {
+function isAbsolute(str: string) : boolean {
```
Comment (jasonLaster) [inline public/js/utils/path.js]: hmm, i could be wrong here, but I think flow can infer the return types

File: public/js/utils/source-map-util.js [hunk 2]
```diff
@@ -1,19 +1,21 @@
...
 
-function generatedToOriginalId(generatedId, url) {
+function generatedToOriginalId(generatedId: string, url: string) : string {
   return generatedId + "/originalSource-" + md5(url);
 }
 
-function isOriginalId(id) {
-  return id.match(/\/originalSource/);
+function isOriginalId(id: string) : boolean {
+  return !!id.match(/\/originalSource/);
```
Comment (jasonLaster) [inline public/js/utils/source-map-util.js]: 👍

File: public/js/types.js [hunk 3]
```diff
@@ -13,8 +13,8 @@
  */
 export type Source = {
   id: string,
-  url?: string,
-  sourceMapURL?: string
+  url: string,
+  sourceMapURL: string
```
Comment (jasonLaster) [inline public/js/types.js]: 👍

File: public/js/utils/source-map-worker.js [hunk 4]
```diff
@@ -64,11 +75,12 @@ function _setSourceMapRoot(sourceMap, absSourceMapURL, source) {
...
 }
 
-function _getSourceMap(generatedSourceId) {
+function _getSourceMap(generatedSourceId: string)
+    : ?Promise<SourceMapConsumer> {
   return sourceMapRequests.get(generatedSourceId);
 }
 
-async function _resolveAndFetch(generatedSource) {
+async function _resolveAndFetch(generatedSource: Source) : SourceMapConsumer {
```
Comment (jasonLaster) [inline public/js/utils/source-map-worker.js]: where did `SourceMapConsumer` return type come from?

**Comment by jasonLaster** [comment]
I just refactored _resolveSourceMapURL so that flow would be happy. The issue there was that flow could not resolve url.parse's properties to strings. Adding the default empty strings did the trick. The rest was just cleanup :)

File: public/js/utils/path.js (hunk 1 above)
Comment (jlongster) [inline public/js/utils/path.js]: If it's an exported function, you need to fully annotate it.

File: public/js/types.js (hunk 3 above)
Comment (jlongster) [inline public/js/types.js]: Why these changes? This is wrong. A source might not have a URL, and definitely will not have a sourceMapURL most of the time.

File: public/js/utils/source-map-worker.js [hunk 5]
```diff
@@ -22,21 +33,21 @@ function enableSourceMaps() {
   sourceMapsEnabled = true;
 }
 
-function _resolveSourceMapURL(source) {
-  if (path.isURL(source.sourceMapURL) || !source.url) {
+function _resolveSourceMapURL(source: Source) : string {
+  const { url, sourceMapURL } = source;
```
Comment (jlongster) [inline public/js/utils/source-map-worker.js]: Please don't destructure this, the code is clearer by simply referencing `source` below. Unless you did this for a specific flow reason?

Comment (jasonLaster) [inline public/js/utils/source-map-worker.js]: will revert

**Comment by jasonLaster** [comment]
@sankha93 i removed the return types as that did not affect the coverage

File: public/js/types.js (hunk 3 above)
Comment (jasonLaster) [inline public/js/types.js]: nice catch, was thinking it would be an empty string, but if the field is missing it should be `?`

Comment (wldcordeiro) [inline public/js/types.js]: If it's nullable the `?` is what you use.

Comment (jasonLaster) [inline public/js/types.js]: yup

File: public/js/utils/source-map-worker.js (hunk 4 above)
Comment (wldcordeiro) [inline public/js/utils/source-map-worker.js]: Seems like this @jasonLaster  https://github.com/mozilla/source-map

**Comment by jasonLaster** [comment]
@sankha93 i made a couple of cleanups based on the comments. It looks like flow is complaining in some cases now because it can't assume sourceMapURL is a string. We will likely need to set reasonable defaults like in URL.parse. Thanks for the help again.

File: public/js/utils/source-map-worker.js [hunk 6]
```diff
@@ -22,21 +33,21 @@ function enableSourceMaps() {
   sourceMapsEnabled = true;
 }
 
-function _resolveSourceMapURL(source) {
-  if (path.isURL(source.sourceMapURL) || !source.url) {
+function _resolveSourceMapURL(source: Source) {
+  const { url, sourceMapURL } = source;
```
Comment (jlongster) [inline public/js/utils/source-map-worker.js]: Just a reminder: (since other comment was on an old commit, I believe) don't destructure here and follow what the code did before.

Comment (sankha93) [inline public/js/utils/source-map-worker.js]: @jlongster I think if we should destructure the `source` object and set the `url` and `sourceMapURL` values to an empty string by default. It would pass Flow's static analysis for `path.isURL` or `path.isAbsolute` since they only expect a string here. /cc @jasonLaster

Comment (sankha93) [inline public/js/utils/source-map-worker.js]: I just pushed the above changes in the latest commit. I'll change it whatever is recommended otherwise.

Comment (jasonLaster) [inline public/js/utils/source-map-worker.js]: Thanks. I think the decision was to avoid destructuring, but still use defaults below.

File: public/js/utils/source-map-worker.js [hunk 7]
```diff
@@ -27,21 +38,21 @@ function enableSourceMaps() {
   sourceMapsEnabled = true;
 }
 
-function _resolveSourceMapURL(source) {
-  if (path.isURL(source.sourceMapURL) || !source.url) {
+function _resolveSourceMapURL(source: Source) {
+  const { url = "", sourceMapURL = "" } = source;
```
Comment (jasonLaster) [inline public/js/utils/source-map-worker.js]: @jlongster @sankha93 I didn't realize that we needed to use the defaults to fix the source type. 👍 

This looks good

**Comment by jasonLaster** [comment]
Thanks for the help @sankha93. I just did a quick check and these fixes brought our score from 42% to 72% in the worker, with obviously no errors!
